### PR TITLE
Add implementation for recover client on message server.

### PR DESCRIPTION
**Pull request by DimitrisJim**
## Description

closes: #4500 

requires #4499 

### Commit Message / Changelog Entry

```bash
type: commit message
```

see the [guidelines](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) for commit messages. (view raw markdown for examples)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/docs/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

File: modules/core/keeper/msg_server_test.go [hunk 1]
```diff
@@ -779,6 +779,94 @@ func (suite *KeeperTestSuite) TestUpgradeClient() {
...
+
+			tc.malleate()
+
+			_, err = keeper.Keeper.RecoverClient(*suite.chainA.App.GetIBCKeeper(), suite.chainA.GetContext(), msg)
+
+			expPass := tc.expErr == nil
+			if expPass {
+				suite.Require().NoError(err)
+
+				// Assert that client status is now Active
```
Comment (DimitrisJim) [inline modules/core/keeper/msg_server_test.go:858]: Similarly as was done for core handler, lmk if some other post recovery state would be more meaningful to check

File: modules/core/keeper/msg_server_test.go [hunk 2]
```diff
@@ -779,6 +779,94 @@ func (suite *KeeperTestSuite) TestUpgradeClient() {
...
+			err := substitutePath.EndpointA.UpdateClient()
+			suite.Require().NoError(err)
+			err = substitutePath.EndpointA.UpdateClient()
+			suite.Require().NoError(err)
+			substituteClientState = suite.chainA.GetClientState(substitute)
+
+			tmClientState, ok := subjectClientState.(*ibctm.ClientState)
+			suite.Require().True(ok)
+			tmClientState.AllowUpdateAfterMisbehaviour = true
+			tmClientState.AllowUpdateAfterExpiry = true
```
Comment (colin-axner) [inline modules/core/keeper/msg_server_test.go:839]: these fields are deprecated, we can remove these lines as they don't produce any change in behaviour

File: modules/core/keeper/msg_server_test.go [hunk 3]
```diff
@@ -779,6 +779,94 @@ func (suite *KeeperTestSuite) TestUpgradeClient() {
...
+			tmClientState.FrozenHeight = tmClientState.LatestHeight
+			suite.chainA.App.GetIBCKeeper().ClientKeeper.SetClientState(suite.chainA.GetContext(), subject, tmClientState)
+
+			tmClientState, ok = substituteClientState.(*ibctm.ClientState)
+			suite.Require().True(ok)
+			tmClientState.AllowUpdateAfterMisbehaviour = true
+			tmClientState.AllowUpdateAfterExpiry = true
```
Comment (colin-axner) [inline modules/core/keeper/msg_server_test.go:846]: ditto

**Review by colin-axner — APPROVED** [review]
Sweet! :sunglasses:

File: modules/core/keeper/msg_server_test.go (hunk 2 above)
Comment (DimitrisJim) [reply modules/core/keeper/msg_server_test.go]: ugh and I knew this

File: modules/core/keeper/msg_server_test.go [hunk 4]
```diff
@@ -779,6 +779,86 @@ func (suite *KeeperTestSuite) TestUpgradeClient() {
 	}
 }
 
+func (suite *KeeperTestSuite) TestRecoverClient() {
+	var (
+		subject, substitute string
+		subjectClientState  exported.ClientState
```
Comment (damiannolan) [inline modules/core/keeper/msg_server_test.go:785]: are these variables needed to be defined here? doesn't look like they're mutated in `malleate()`

**Review by damiannolan — APPROVED** [review]
Excellent stuff! 🚀

File: modules/core/keeper/msg_server_test.go (hunk 4 above)
Comment (DimitrisJim) [reply modules/core/keeper/msg_server_test.go]: nice catch, yea can be axed from the var block

**Pull request merged**